src/native/libs/System.Globalization.Native/pal_collation.m

```diff
@@ -45,33 +60,215 @@ static NSStringCompareOptions ConvertFromCompareOptionsToNSStringCompareOptions(
 Function:
 CompareString
 */
-int32_t GlobalizationNative_CompareStringNative(const uint16_t* localeName, int32_t lNameLength, const uint16_t* lpStr1, int32_t cwStr1Length, 
-                                                const uint16_t* lpStr2, int32_t cwStr2Length, int32_t comparisonOptions)
+int32_t GlobalizationNative_CompareStringNative(const uint16_t* localeName, int32_t lNameLength, const uint16_t* lpSource, int32_t cwSourceLength, 
+                                                const uint16_t* lpTarget, int32_t cwTargetLength, int32_t comparisonOptions)
 {    
-    NSLocale *currentLocale;
-    if(localeName == NULL || lNameLength == 0)
+    NSLocale *currentLocale = GetCurrentLocale(localeName, lNameLength);
+    NSString *sourceString = [NSString stringWithCharacters: lpSource length: cwSourceLength];
+    NSString *sourceStrComposed = sourceString.precomposedStringWithCanonicalMapping;
+    NSString *targetString = [NSString stringWithCharacters: lpTarget length: cwTargetLength];
+    NSString *targetStrComposed = targetString.precomposedStringWithCanonicalMapping;
+
+    NSRange comparisonRange = NSMakeRange(0, sourceStrComposed.length);
+    NSStringCompareOptions options = ConvertFromCompareOptionsToNSStringCompareOptions(comparisonOptions);
+    
+    // in case mapping is not found
+    if (options == 0)
+        return -2;
+
+    return [sourceStrComposed compare:targetStrComposed
+                              options:options
+                              range:comparisonRange
+                              locale:currentLocale];
+}
+
+static NSString* RemoveWeightlessCharacters(NSString* source)
+{
+    NSError *error = nil;
+    NSRegularExpression *regex = [NSRegularExpression regularExpressionWithPattern:@"[\u200B-\u200D\uFEFF\0]" options:NSRegularExpressionCaseInsensitive error:&error];
+
+    if (error != nil)
+        return source;
+
+    NSString *modifiedString = [regex stringByReplacingMatchesInString:source options:0 range:NSMakeRange(0, [source length]) withTemplate:@""];
+
+    return modifiedString;
+}
+
+// Remove weightless characters and normalize string with form C
+static NSString* ComposeString(NSString* source)
+{
+    return RemoveWeightlessCharacters(source.precomposedStringWithCanonicalMapping);
+}
+
+/*
+Function: IndexOf
+Find detailed explanation how this function works in https://github.com/dotnet/runtime/blob/main/docs/design/features/globalization-hybrid-mode.md
```

Comment:
I'm still not completely sure whats happening in this function even after trying to parse the different cases in `globalization-hybrid-mode.md`. Could it be elaborated why all of these "fallbacks" will cover all of the cases (how do we know we are not missing another fallback)?

I think Form C and D can be elaborated further in the readme, or atleast a link to where they will be explained.
How come we are only using canonical equivalence in Normalization Form C (NFC) and Normalization Form D (NFD) and not using any of the compatibility equivalence in Normalization Form KC (NFKC) and Normalization Form KD (NFKD)?

From what I can tell, there are a bunch of early return conditions in which we wouldn't use any apple apis or even the Form C/Form D logic, like
1. if either string is empty
2. options not supported (maybe this should be first)
3. Using an API to compare the strings without weightless characters in Normalization Form C (what is the rationale for using NFC here for an early check?) We don't seem to use `sourceStrPrecomposed` anywhere afterwards, so why does the follow logic will comprehensively cover the scenario in which `sourceStrPrecomposed` contains `searchStrPrecomposed`?

`localizedStandardRangeOfString` seems extremely specific, why does the following logic cover all scenarios in which `(comparisonOptions & IgnoreCase) && lNameLength == 0 && fromBeginning` is `false`, but not this particular case?

After that, is it just that we are comparing 
sourceStrComposed vs. searchStrComposed
sourceStrComposed vs. searchStrPrecomposed
sourceStrComposed vs. searchStrDecomposed?
The if else for the latter two makes the whole flow a bit confusing, can we maintain parallelism by changing
```
    rangeOfReceiverToSearch = NSMakeRange(0, sourceStrComposed.length);
    // Normalize search string with Form C
    NSRange preComposedRange = [sourceStrComposed rangeOfString:searchStrPrecomposed
                                                  options:options
                                                  range:rangeOfReceiverToSearch
                                                  locale:currentLocale];

    if (preComposedRange.location != NSNotFound)
    {
        // in case of last index and CompareOptions.IgnoreCase 
        // if letters have different representations in source and search strings
        // and search appears more than one time in source string take last index
        // e.g. new CultureInfo().CompareInfo.LastIndexOf("Is \u0055\u0308 or \u0075\u0308 the same as \u00DC or \u00FC?", "U\u0308", 25,18, CompareOptions.IgnoreCase);
        // this will return 24 
        if ((int32_t)result.location > (int32_t)preComposedRange.location && !fromBeginning && (comparisonOptions & IgnoreCase))
            return result;
        result.location = preComposedRange.location;
        result.length = preComposedRange.length;
    }
    else
    {
        // Normalize search string with Form D
        NSString *searchStrDecomposed = searchStrComposed.decomposedStringWithCanonicalMapping;
        NSRange deComposedRange = [sourceStrComposed rangeOfString:searchStrDecomposed
                                                     options:options
                                                     range:rangeOfReceiverToSearch
                                                     locale:currentLocale];

        if (deComposedRange.location != NSNotFound)
        {
            result.location = deComposedRange.location;
            result.length = deComposedRange.length;                    
            return result;
        }
    }

    return result;
```
to
```
    rangeOfReceiverToSearch = NSMakeRange(0, sourceStrComposed.length);
    // Normalize search string with Form C
    NSRange preComposedRange = [sourceStrComposed rangeOfString:searchStrPrecomposed
                                                  options:options
                                                  range:rangeOfReceiverToSearch
                                                  locale:currentLocale];

    if (preComposedRange.location != NSNotFound)
    {
        // in case of last index and CompareOptions.IgnoreCase 
        // if letters have different representations in source and search strings
        // and search appears more than one time in source string take last index
        // e.g. new CultureInfo().CompareInfo.LastIndexOf("Is \u0055\u0308 or \u0075\u0308 the same as \u00DC or \u00FC?", "U\u0308", 25,18, CompareOptions.IgnoreCase);
        // this will return 24 
        if ((int32_t)result.location > (int32_t)preComposedRange.location && !fromBeginning && (comparisonOptions & IgnoreCase))
            return result;
        result.location = preComposedRange.location;
        result.length = preComposedRange.length;
        return result;
    }
        
    // Normalize search string with Form D
    NSString *searchStrDecomposed = searchStrComposed.decomposedStringWithCanonicalMapping;
    NSRange deComposedRange = [sourceStrComposed rangeOfString:searchStrDecomposed
                                                 options:options
                                                 range:rangeOfReceiverToSearch
                                                 locale:currentLocale];
    if (deComposedRange.location != NSNotFound)
    {
        result.location = deComposedRange.location;
        result.length = deComposedRange.length;                    
        return result;
    }

    return result; // Maybe we should throw or assert that we shouldn't get here if everything above should cover everything?
```

Reply:
1. As for normalisation forms I am only checking` D (NFD)` and `C (NFC)` inside `rangeOfString:options:range:locale: `function as `NSString` objects are compared by checking the Unicode `canonical equivalence` of their code point sequences (see https://developer.apple.com/documentation/foundation/nsstring/1417348-rangeofstring) so only canonical mapping is needed. Normalization Form KC (NFKC) and Normalization Form KD (NFKD) are for compatibility equivalence.
2. `localizedStandardRangeOfString`  is performing case and diacritic insensitive, locale-aware search by implementation, that is why I am checking for `IgnoreCase` and if locale is not passed .(see https://developer.apple.com/documentation/foundation/nsstring/1413574-localizedstandardrangeofstring)

Reply:
Ah, I think I am understanding this more.

Characters in general are represented by unicode code points, and some characters can be represented in a single code point or by combining multiple characters (like diacritics/diaeresis). Normalization Form C will look to compress characters to their single code point format if they were originally represented as a sequence of multiple code points. Normalization Form D does the opposite and expands characters into their multiple code point formats if possible.

`NSString` `rangeOfString:options:range:locale:` uses canonical equivalence to find the position of the `searchString` within the `sourceString`, however, it does not automatically handle comparison of precomposed (single code point representation) or decomposed (most code points representation). Because the `searchString` and `sourceString` can be of differing formats, to properly find the index, we need to ensure that the `searchString` is in the same form as the `sourceString` by checking the `rangeOfString:options:range:locale:` using every single normalization form.

Is my understanding correct?
I think we can elaborate on the rationale in the readme to reflect something like the above.


---------------------

```
    // check if source contains search string
    rangeOfReceiverToSearch = NSMakeRange(0, sourceStrPrecomposed.length);
    NSRange containsRange = [sourceStrPrecomposed rangeOfString:searchStrPrecomposed
                                                  options:options
                                                  range:rangeOfReceiverToSearch
                                                  locale:currentLocale];
```
Can we expand this to show that we don't care which form, but by normalizing both strings to same form, we can see if there is a possible match, making it worth to check every form.

I think the ordering in which you have
```
sourceStrComposed vs. searchStrComposed
sourceStrComposed vs. searchStrPrecomposed
sourceStrComposed vs. searchStrDecomposed
```
 is best.
 
 Can we have a comment like
 
```
// sourceString and searchString possibly have the same composition of characters
sourceStrComposed vs. searchStrComposed

// maybe sourceString has precomposed form of characters and searchString has decomposed form of characters
// convert searchString to a precomposed form
sourceStrComposed vs. searchStrPrecomposed

// maybe sourceString has decomposed form of characters and searchString has precomposed form of characters
// convert searchString to a decomposed form
sourceStrComposed vs. searchStrDecomposed?
```

Reply:
Yes, that will make the code more readable. Updated.